File: backend/services/pozos_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from datetime import datetime
from typing import Optional

from repositories.pozos_repository import PozosRepository
from schemas.pozos import PozoItem, SerieTemporal, PuntoTemporal, ResumenPozo

UNIDADES = {
    "caudal":      "m³/h",
    "nivel":       "m",
    "totalizador": "",
}


class PozosService:

    def __init__(self, db: AsyncSession):
        self.repo = PozosRepository(db)

    async def listar_pozos(self) -> list[PozoItem]:
        rows = await self.repo.listar_pozos()
        return [
            PozoItem(
                entity_id=r["entity_id"],
                ultimo_registro=r.get("ultimo_registro"),
                total_registros=r.get("total_registros"),
            )
            for r in rows
        ]

    async def serie_temporal(
        self,
        entity_id: str,
        variable: str,
        desde: Optional[datetime],
        hasta: Optional[datetime],
        intervalo: str = "1h",
    ) -> SerieTemporal:
        rows = await self.repo.serie_temporal(entity_id, variable, desde, hasta, intervalo)
        datos = [
            PuntoTemporal(
                tiempo=r["tiempo"],
                valor=round(float(r["valor"]), 4) if r.get("valor") is not None else None,
            )
            for r in rows
        ]
        return SerieTemporal(
            entity_id=entity_id,
            variable=variable,
            unidad=UNIDADES.get(variable, ""),
            desde=desde,
            hasta=hasta,
            intervalo=intervalo,
            datos=datos,
        )

    async def resumen_pozo(
        self,
        entity_id: str,
        desde: Optional[datetime],
        hasta: Optional[datetime],
    ) -> ResumenPozo:
        r = await self.repo.resumen_pozo(entity_id, desde, hasta)

        def _round(v) -> Optional[float]:
            return round(float(v), 4) if v is not None else None

        return ResumenPozo(
            entity_id=entity_id,
            caudal_promedio=_round(r.get("caudal_promedio")),
            caudal_maximo=_round(r.get("caudal_maximo")),
            caudal_minimo=_round(r.get("caudal_minimo")),
            nivel_promedio=_round(r.get("nivel_promedio")),
            nivel_maximo=_round(r.get("nivel_maximo")),
            nivel_minimo=_round(r.get("nivel_minimo")),
            totalizador_ultimo=_round(r.get("totalizador_ultimo")),
            ultimo_registro=r.get("ultimo_registro"),
            total_registros=r.get("total_registros", 0),
        )

File: frontend/services/pozos.service.ts
// Servicio para consumir la API de Pozos (TimescaleDB)
const API_BASE = "";



// ── Tipos ─────────────────────────────────────────────────────────────────────

export interface PozoItem {
  entity_id: string;
  ultimo_registro: string | null;
  total_registros: number | null;
}

export interface PuntoTemporal {
  tiempo: string;  // ISO 8601
  valor: number | null;
}

export interface SerieTemporal {
  entity_id: string;
  variable: string;
  unidad: string;
  desde: string | null;
  hasta: string | null;
  intervalo: string;
  datos: PuntoTemporal[];
}

export interface ResumenPozo {
  entity_id: string;
  caudal_promedio: number | null;
  caudal_maximo: number | null;
  caudal_minimo: number | null;
  nivel_promedio: number | null;
  nivel_maximo: number | null;
  nivel_minimo: number | null;
  totalizador_ultimo: number | null;
  ultimo_registro: string | null;
  total_registros: number;
}

export type Intervalo = "5min" | "15min" | "1h" | "6h" | "1day";
export type Variable = "caudal" | "nivel" | "totalizador";

// ── Helpers ───────────────────────────────────────────────────────────────────

function buildParams(params: Record<string, string | undefined | null>): string {
  const sp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v != null && v !== "") sp.append(k, v);
  }
  const str = sp.toString();
  return str ? `?${str}` : "";
}

async function apiFetch<T>(path: string): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    cache: "no-store",
    headers: { "Content-Type": "application/json" },
  });
  if (!res.ok) {
    const detail = await res.text();
    throw new Error(`API error ${res.status}: ${detail}`);
  }
  return res.json() as Promise<T>;
}

// ── Funciones públicas ────────────────────────────────────────────────────────

/**
 * Lista todos los pozos únicos (entity_ids) con su último registro.
 */
export async function fetchPozos(): Promise<PozoItem[]> {
  return apiFetch<PozoItem[]>("/api/v1/pozos");
}

/**
 * Obtiene la serie temporal de CAUDAL (qc) para un pozo.
 */
export async function fetchCaudal(
  entityId: string,
  desde?: string,
  hasta?: string,
  intervalo: Intervalo = "1h"
): Promise<SerieTemporal> {
  const qs = buildParams({ desde, hasta, intervalo });
  return apiFetch<SerieTemporal>(`/api/v1/pozos/${entityId}/caudal${qs}`);
}

/**
 * Obtiene la serie temporal de NIVEL FREÁTICO (nc) para un pozo.
 */
export async function fetchNivel(
  entityId: string,
  desde?: string,
  hasta?: string,
  intervalo: Intervalo = "1h"
): Promise<SerieTemporal> {
  const qs = buildParams({ desde, hasta, intervalo });
  return apiFetch<SerieTemporal>(`/api/v1/pozos/${entityId}/nivel${qs}`);
}

/**
 * Obtiene la serie temporal del TOTALIZADOR (tc) para un pozo.
 */
export async function fetchTotalizador(
  entityId: string,
  desde?: string,
  hasta?: string,
  intervalo: Intervalo = "1h"
): Promise<SerieTemporal> {
  const qs = buildParams({ desde, hasta, intervalo });
  return apiFetch<SerieTemporal>(`/api/v1/pozos/${entityId}/totalizador${qs}`);
}

/**
 * Obtiene el resumen / KPIs de un pozo en un rango de fechas.
 */
export async function fetchResumenPozo(
  entityId: string,
  desde?: string,
  hasta?: string
): Promise<ResumenPozo> {
  const qs = buildParams({ desde, hasta });
  return apiFetch<ResumenPozo>(`/api/v1/pozos/${entityId}/resumen${qs}`);
}

/**
 * Carga en paralelo el caudal, nivel y resumen de un pozo.
 * Útil para cargar todos los datos del gráfico de una vez.
 */
export async function fetchDatosPozo(
  entityId: string,
  desde?: string,
  hasta?: string,
  intervalo: Intervalo = "1h"
): Promise<{
  caudal: SerieTemporal;
  nivel: SerieTemporal;
  resumen: ResumenPozo;
}> {
  const [caudal, nivel, resumen] = await Promise.all([
    fetchCaudal(entityId, desde, hasta, intervalo),
    fetchNivel(entityId, desde, hasta, intervalo),
    fetchResumenPozo(entityId, desde, hasta),
  ]);
  return { caudal, nivel, resumen };
}
